### ISSUE TITLE
ICE when matching on enum variant constructors as fn pointers

### ISSUE THREAD

**Issue by matklad** [rust-lang/rust#48205]
```Rust
enum Variant<T> {
    A(T),
    B(T),
}

fn main() {
    const A: fn(usize) -> Variant<usize> = Variant::A;
    const B: fn(usize) -> Variant<usize> = Variant::B;

    let my_variant: fn(usize) -> Variant<usize> = Variant::A;

    match my_variant {
        A => {}
        B => {}
        _ => {}
    }
}
```

```
error: internal compiler error: broken MIR in NodeId(13) (_3 = discriminant(_1)): bad assignment (isize = u8): Sorts(ExpectedFound { expected: isize, found: u8 })
  --> src/main.rs:13:9
   |
13 |         A => {}
   |         ^

Switch constants must all be same type as switch value!
  switch i8 0, label %5 [
    i64 0, label %3
    i64 1, label %4
  ], !dbg !36
LLVM ERROR: Broken function found, compilation aborted!
```

[playground](https://play.rust-lang.org/?gist=f3e3ebcc6928656d08894940c4bca5ca&version=nightly)

Original example by @chessnokov at https://gitter.im/ruRust/general?at=5a84190c5cc1872645398cd6

**Comment by Aaronepower** [comment]
It seems this has been fixed in beta(`1.26.0-beta.11 2018-04-23 5181002c66af49bf73a2`).

**Comment by matklad** [comment]
Confirmed that this now works!
